[WTF] Align Windows OSAllocator behavior more to the other platforms

Constellation opened the pull request:
#### b0b1cd7c9eae5f100fc535c0475b140f36ab38d7
<pre>
[WTF] Align Windows OSAllocator behavior more to the other platforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=243050">https://bugs.webkit.org/show_bug.cgi?id=243050</a>

Reviewed by Fujii Hironori.

Use VirtualAlloc + MEM_RESET, which is more aligned to madvise + MADV_FREE,
and it is more efficient than decommitting synchronously.

* Source/WTF/wtf/win/OSAllocatorWin.cpp:
(WTF::OSAllocator::decommit):

Canonical link: <a href="https://commits.webkit.org/252694@main">https://commits.webkit.org/252694@main</a>
</pre>

fujii commented on Source/WTF/wtf/win/OSAllocatorWin.cpp:
```diff
@@ -97,19 +97,21 @@ void OSAllocator::commit(void* address, size_t bytes, bool writable, bool execut
+    // https://devblogs.microsoft.com/oldnewthing/20170113-00/?p=95185
+    // > The fact that MEM_RESET does not remove the page from the working set is not actually mentioned
+    // > in the documentation for the MEM_RESET flag. Instead, it’s mentioned in the documentation for
+    // > the Offer­Virtual­Memory function, and in a sort of backhanded w
```
backhanded w → backhanded way

Constellation replied:
Fixed.

fujii commented on Source/WTF/wtf/win/OSAllocatorWin.cpp:
```diff
@@ -97,19 +97,21 @@ void OSAllocator::commit(void* address, size_t bytes, bool writable, bool execut
-    // Silence warning about using MEM_DECOMMIT instead of MEM_RELEASE:
-#pragma warning(suppress: 6250)
-    bool result = VirtualFree(address, bytes, MEM_DECOMMIT);
+    void* result = VirtualAlloc(address, bytes, MEM_RESET, PAGE_READWRITE);
```
I think this is a behavior change. committing again by using `OSAllocator::commit` returned not zero-filled region. Is it OK?

Constellation replied:
Yes, it works. And the other platforms' OSAllocator does not guarantee that zero-filled pages are produced from commit. So, the behavior is more aligned to the other platforms' one.

fujii commented on Source/WTF/wtf/win/OSAllocatorWin.cpp:
```diff
@@ -97,19 +97,21 @@ void OSAllocator::commit(void* address, size_t bytes, bool writable, bool execut
     if (!result)
         CRASH();
+    if (result == address)
+        VirtualUnlock(address, bytes);
```
Is it OK to use `VirtualUnlock` without `VirtualLock`?

Constellation replied:
Yes, this is allowed. https://devblogs.microsoft.com/oldnewthing/20170317-00/?p=95755
I've attached comment with the above URL.

fujii replied:
The official document of VirtualUnlock says it. This looks a bit tricky code. We should add more comment and the URL of the official document rather than the blog URL. For example,
> // Calling VirtualUnlock on a range of memory that is not locked releases the pages from the process's working set.
> // https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualunlock

Constellation replied:
Thanks, fixed.

fujii commented on Source/WTF/wtf/win/OSAllocatorWin.cpp:
```diff
@@ -97,19 +97,22 @@ void OSAllocator::commit(void* address, size_t bytes, bool writable, bool execut
     if (!result)
         CRASH();
+    // https://devblogs.microsoft.com/oldnewthing/20170317-00/?p=95755
+    if (result == address)
```
Is this condition needed?

Constellation replied:
This is not necessary. Removed.

Constellation committed:
[WTF] Align Windows OSAllocator behavior more to the other platforms
https://bugs.webkit.org/show_bug.cgi?id=243050

Reviewed by Fujii Hironori.

Use VirtualAlloc + MEM_RESET, which is more aligned to madvise + MADV_FREE,
and it is more efficient than decommitting synchronously.

* Source/WTF/wtf/win/OSAllocatorWin.cpp:
(WTF::OSAllocator::decommit):

Canonical link: https://commits.webkit.org/252694@main
